### ISSUE TITLE
wrong type annotations do not always throw an error

### ISSUE THREAD

**Issue by amtoine** [nushell/nushell#9598]
### Describe the bug

today i learned we can annotate variable declarations with type information :thumbsup:
however, i think the type checking is a bit too weak here :yum:

### How to reproduce

1. let's define the two following commands
```nushell
> def cmd_int [] { return 42 }
```
```nushell
> def cmd_string [] { return "my string" }
```
2. the following works fine
```nushell
> let x: int = cmd_int; $x
42
> let y: string = cmd_string; $y
my string
```
3. however the following works as well...
```
> let x: string = cmd_int; $x
42
> let y: int = cmd_string; $y
my string
```

### Expected behavior

i expected something more like
```nushell
> let x: string = cmd_int; $x
Error: nu::parser::type_mismatch

  × Type mismatch.
   ╭─[entry #1:1:1]
 1 │ let x: string = cmd_int; $x
   ·                 ───┬───
   ·                    ╰── expected string, found int
   ╰────
```
and
```nushell
> let y: int = cmd_string; $y
Error: nu::parser::type_mismatch

  × Type mismatch.
   ╭─[entry #1:1:1]
 1 │ let y: int = cmd_string; $y
   ·              ─────┬────
   ·                   ╰── expected int, found string
   ╰────
```

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.82.1                                   |
| branch             | main                                     |
| commit_hash        | 2bb0c1c618f961843b49432fb7a21304b41493af |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.68.2 (9eb3afe9e 2023-03-27)      |
| rust_channel       | 1.68.2-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.68.2 (6feb7c9cf 2023-03-26)      |
| build_time         | 2023-07-03 18:32:11 +02:00               |
| build_rust_channel | release                                  |
| allocator          | standard                                 |
| features           | default, sqlite, trash, which, zip       |
| installed_plugins  | hist, plot                               |

**Comment by amtoine** [comment]
amazing @1Kinoti :blush:

**Comment by 1Kinoti** [comment]
@amtoine upon further investigation, [this comment](https://github.com/nushell/nushell/issues/9598#issuecomment-1623632331) is WRONG. 

the issue isn't type annotations not being checked because [they are checked](https://github.com/nushell/nushell/blob/667502e8da0ef8bc06c10f62684c76debae4cf50/crates/nu-parser/src/parse_keywords.rs#L2857)

the issue is specific to `string`s, and is related to [this](https://github.com/nushell/nushell/pull/8635) and [this](https://github.com/nushell/nushell/pull/8845). 

tldr; `int`s are 'implicitly cast' into `string`s, ie, where `string`s are expected, `int` will parse

**Comment by amtoine** [comment]
thanks for the further inspection of what happens @1Kinoti :pray: 
are you thinking of a patch for this? :innocent:

**Comment by 1Kinoti** [comment]
i think it works as expected, if we had a way to separate the shell from the file during parsing, then we'd be able to make parsing a file more stricter.

the implicit cast is imo the right behavior in the shell side of things, having to quote strings that look like numbers in a command line is annoying.

@jntrnr had introduced this strict separation in #8635, but it broke A LOT of people's workflows, and was reverted.

**Comment by jntrnr** [comment]
your two commands have `any` as their output type, so that wouldn't get checked by the typechecker. You need input/output types first.

**Comment by 1Kinoti** [comment]
@amtoine is this now fixed?

**Comment by amtoine** [comment]
@1Kinoti 
> is this now fixed?
>
i've just built latest Nushell and, as far as i can tell, this is not fixed :thinking:

**Comment by 1Kinoti** [comment]
mmmmmh,🤔 even with the function's return types indicated? i'll have do some more investigation then

**Comment by amtoine** [comment]
> mmmmmh,thinking even with the function's return types indicated?
>
ooh i haven't tried with annotations :open_mouth: 

so the types are not infered here, right?

**Comment by amtoine** [comment]
so you were right @1Kinoti :yum:
if i add annotations
```nushell
def cmd_string []: nothing -> string { "my string" }
def cmd_int []: nothing -> int { 123 }
```
the correct type associations still work
```nushell
> let x: int = cmd_int; $x
123
> let s: string = cmd_string; $s
my string
```
and when trying to switch types, i get nice errors
```nushell
> let s: string = cmd_int; $s
Error: nu::parser::type_mismatch

  × Type mismatch.
   ╭─[entry #7:1:1]
 1 │ let s: string = cmd_int; $s
   ·                 ───┬───
   ·                    ╰── expected string, found int
   ╰────
```
and
```nushell
> let x: string = cmd_int; $x
Error: nu::parser::type_mismatch

  × Type mismatch.
   ╭─[entry #8:1:1]
 1 │ let x: string = cmd_int; $x
   ·                 ───┬───
   ·                    ╰── expected string, found int
   ╰────
```

**Comment by 1Kinoti** [comment]
> should the types be infered?

this would be the natural next step, now that block i/o types are known, we should be able to infer the typesbeven without explicit annotations. (although, wouldn't this make nu stricter? not sure)

> otherwise, we probably want to update the book at some point to explain how type annotations and checking works 😋

i think we should be consistent either way. 
```nushell
def x [] { 4 }

# and

def x []: nothing -> int { 4 }
```
should not show different behavior

**Comment by amtoine** [comment]
> i think we should be consistent either way.
> 
> ```nushell
> def x [] { 4 }
> 
> # and
> 
> def x []: nothing -> int { 4 }
> ```
> 
> should not show different behavior
>
unless we want annotations to be required :smirk: 
(agree about the consistency!)

**Comment by Bahex** [comment]
As of `0.114.1` with the `enforce-runtime-annotations` experimental option enabled, assignments are properly type checked:
```nushell
def cmd_int [] { return 42 }
def cmd_string [] { return "my string" }

let x: string = cmd_int; $x
#=> Error: nu::shell::type_mismatch
#=> 
#=>   × Type mismatch.
#=>    ╭─[repl_entry #11:1:25]
#=>  1 │ def cmd_int [] { return 42 }
#=>    ·                         ─┬
#=>    ·                          ╰── the value is a int
#=>  2 │ def cmd_string [] { return "my string" }
#=>  3 │ 
#=>  4 │ let x: string = cmd_int; $x
#=>    · ─┬─
#=>    ·  ╰── expected string, got int
#=>    ╰────
```
